### DISCUSSION TITLE
Transcript cutoff

### DISCUSSION THREAD

**Discussion by aawezk786** [deepgram/community#485] [General help]
Hi Team,
I am using deepgram nova-2 in my nodejs server application (deepgram lib)
Issue:-
Facing transcript cutoff although i have set manual utterence time in deepgram params as well

Deepgram Configuration:-
deepgram.transcription.live({
      language: "en-US",
      punctuate: false,
      encoding: "linear16",
      sample_rate: 16000,
      numerals: true,
      // endpointing: 2000,It also contains boolean value
      utterance_end_ms: 2000, //silence detection timeout
      interim_results: true,
      smart_format: true,
      model: process.env.DEEPGRAM_MODEL,
    });
    
 **Response:-**
 
 **Start Response:-**
 
{"type":"Results","channel_index":[0,1],"duration":3.0100002,"start":11.49,"is_final":false,"speech_final":false,"channel":{"alternatives":[{"transcript":"","confidence":0,"words":[]}]},"metadata":{"request_id":"49397075-7dfb-4848-8419-190bc3d0d9b9","model_info":{"name":"2-ea-nova","version":"2023-09-18.22988","arch":"nova-2"},"model_uuid":"79dd3ac5-776e-4fcf-b4fb-0f9968bac445"}}  
  
  **Final Response received:-**
  
**Final when utterence end event type received({"type":"UtteranceEnd"})**

{"type":"Results","channel_index":[0,1],"duration":3.7199993,"start":18.58,"is_final":true,"speech_final":true,"channel":{"alternatives":[{"transcript":**"This is",**"confidence":0.9850814,"words":[{"word":"this","start":20.98,"end":21.3,"confidence":0.9850814,"punctuated_word":"This"},{"word":"is","start":21.3,"end":21.8,"confidence":0.5317576,"punctuated_word":"is"}]}]},"metadata":{"request_id":"49397075-7dfb-4848-8419-190bc3d0d9b9","model_info":{"name":"2-ea-nova","version":"2023-09-18.22988","arch":"nova-2"},"model_uuid":"79dd3ac5-776e-4fcf-b4fb-0f9968bac445"}}

**Full logs of Transcription:-**

 Transcription ==>  {"type":"Results","channel_index":[0,1],"duration":3.9999995,"start":4.28,"is_final":false,"speech_final":false,"channel":{"alternatives":[{"transcript":"","confidence":0,"words":[]}]},"metadata":{"request_id":"49397075-7dfb-4848-8419-190bc3d0d9b9","model_info":{"name":"2-ea-nova","version":"2023-09-18.22988","arch":"nova-2"},"model_uuid":"79dd3ac5-776e-4fcf-b4fb-0f9968bac445"}}
deepgram: transcript received
socket: transcript sent to client
deepgram: packet received

 Time 2023-12-07T22:47:11.682Z

 Transcription ==>  {"type":"Results","channel_index":[0,1],"duration":4.19,"start":4.28,"is_final":true,"speech_final":false,"channel":{"alternatives":[{"transcript":"","confidence":0,"words":[]}]},"metadata":{"request_id":"49397075-7dfb-4848-8419-190bc3d0d9b9","model_info":{"name":"2-ea-nova","version":"2023-09-18.22988","arch":"nova-2"},"model_uuid":"79dd3ac5-776e-4fcf-b4fb-0f9968bac445"}}
deepgram: transcript received
socket: transcript sent to client
deepgram: packet received

 Time 2023-12-07T22:47:11.840Z

 Transcription ==>  {"type":"Results","channel_index":[0,1],"duration":1.0099993,"start":8.47,"is_final":false,"speech_final":false,"channel":{"alternatives":[{"transcript":"","confidence":0,"words":[]}]},"metadata":{"request_id":"49397075-7dfb-4848-8419-190bc3d0d9b9","model_info":{"name":"2-ea-nova","version":"2023-09-18.22988","arch":"nova-2"},"model_uuid":"79dd3ac5-776e-4fcf-b4fb-0f9968bac445"}}
deepgram: transcript received
socket: transcript sent to client
deepgram: packet received

 Time 2023-12-07T22:47:12.867Z

 Transcription ==>  {"type":"Results","channel_index":[0,1],"duration":2.0099993,"start":8.47,"is_final":false,"speech_final":false,"channel":{"alternatives":[{"transcript":"","confidence":0,"words":[]}]},"metadata":{"request_id":"49397075-7dfb-4848-8419-190bc3d0d9b9","model_info":{"name":"2-ea-nova","version":"2023-09-18.22988","arch":"nova-2"},"model_uuid":"79dd3ac5-776e-4fcf-b4fb-0f9968bac445"}}
deepgram: transcript received
socket: transcript sent to client
deepgram: packet received

 Time 2023-12-07T22:47:13.367Z

 Transcription ==>  {"type":"Results","channel_index":[0,1],"duration":2.46,"start":8.47,"is_final":true,"speech_final":true,"channel":{"alternatives":[{"transcript":"","confidence":0,"words":[]}]},"metadata":{"request_id":"49397075-7dfb-4848-8419-190bc3d0d9b9","model_info":{"name":"2-ea-nova","version":"2023-09-18.22988","arch":"nova-2"},"model_uuid":"79dd3ac5-776e-4fcf-b4fb-0f9968bac445"}}
deepgram: transcript received
socket: transcript sent to client
deepgram: packet received

 Time 2023-12-07T22:47:13.876Z

 Transcription ==>  {"type":"Results","channel_index":[0,1],"duration":0.55999947,"start":10.93,"is_final":true,"speech_final":true,"channel":{"alternatives":[{"transcript":"","confidence":0,"words":[]}]},"metadata":{"request_id":"49397075-7dfb-4848-8419-190bc3d0d9b9","model_info":{"name":"2-ea-nova","version":"2023-09-18.22988","arch":"nova-2"},"model_uuid":"79dd3ac5-776e-4fcf-b4fb-0f9968bac445"}}
deepgram: transcript received
socket: transcript sent to client
deepgram: packet received

 Time 2023-12-07T22:47:14.867Z

 Transcription ==>  {"type":"Results","channel_index":[0,1],"duration":1.0100002,"start":11.49,"is_final":false,"speech_final":false,"channel":{"alternatives":[{"transcript":"","confidence":0,"words":[]}]},"metadata":{"request_id":"49397075-7dfb-4848-8419-190bc3d0d9b9","model_info":{"name":"2-ea-nova","version":"2023-09-18.22988","arch":"nova-2"},"model_uuid":"79dd3ac5-776e-4fcf-b4fb-0f9968bac445"}}
deepgram: transcript received
socket: transcript sent to client
deepgram: packet received

 Time 2023-12-07T22:47:15.904Z

 Transcription ==>  {"type":"Results","channel_index":[0,1],"duration":2.0100002,"start":11.49,"is_final":false,"speech_final":false,"channel":{"alternatives":[{"transcript":"","confidence":0,"words":[]}]},"metadata":{"request_id":"49397075-7dfb-4848-8419-190bc3d0d9b9","model_info":{"name":"2-ea-nova","version":"2023-09-18.22988","arch":"nova-2"},"model_uuid":"79dd3ac5-776e-4fcf-b4fb-0f9968bac445"}}
deepgram: transcript received
socket: transcript sent to client
deepgram: packet received

 Time 2023-12-07T22:47:15.921Z

 Transcription ==>  {"type":"UtteranceEnd"}
deepgram: utterence end

 Time 2023-12-07T22:47:15.921Z
deepgram: packet received

 Time 2023-12-07T22:47:16.887Z

 Transcription ==>  {"type":"Results","channel_index":[0,1],"duration":3.0100002,"start":11.49,"is_final":false,"speech_final":false,"channel":{"alternatives":[{"transcript":"","confidence":0,"words":[]}]},"metadata":{"request_id":"49397075-7dfb-4848-8419-190bc3d0d9b9","model_info":{"name":"2-ea-nova","version":"2023-09-18.22988","arch":"nova-2"},"model_uuid":"79dd3ac5-776e-4fcf-b4fb-0f9968bac445"}}
deepgram: transcript received
socket: transcript sent to client
deepgram: packet received

 Time 2023-12-07T22:47:17.887Z

 Transcription ==>  {"type":"Results","channel_index":[0,1],"duration":4.01,"start":11.49,"is_final":false,"speech_final":false,"channel":{"alternatives":[{"transcript":"","confidence":0,"words":[]}]},"metadata":{"request_id":"49397075-7dfb-4848-8419-190bc3d0d9b9","model_info":{"name":"2-ea-nova","version":"2023-09-18.22988","arch":"nova-2"},"model_uuid":"79dd3ac5-776e-4fcf-b4fb-0f9968bac445"}}
deepgram: transcript received
socket: transcript sent to client
deepgram: packet received

 Time 2023-12-07T22:47:18.934Z

 Transcription ==>  {"type":"Results","channel_index":[0,1],"duration":3.9499998,"start":11.49,"is_final":true,"speech_final":false,"channel":{"alternatives":[{"transcript":"","confidence":0,"words":[]}]},"metadata":{"request_id":"49397075-7dfb-4848-8419-190bc3d0d9b9","model_info":{"name":"2-ea-nova","version":"2023-09-18.22988","arch":"nova-2"},"model_uuid":"79dd3ac5-776e-4fcf-b4fb-0f9968bac445"}}
deepgram: transcript received
socket: transcript sent to client
deepgram: packet received

 Time 2023-12-07T22:47:18.940Z

 Transcription ==>  {"type":"Results","channel_index":[0,1],"duration":1.0600004,"start":15.44,"is_final":false,"speech_final":false,"channel":{"alternatives":[{"transcript":"","confidence":0,"words":[]}]},"metadata":{"request_id":"49397075-7dfb-4848-8419-190bc3d0d9b9","model_info":{"name":"2-ea-nova","version":"2023-09-18.22988","arch":"nova-2"},"model_uuid":"79dd3ac5-776e-4fcf-b4fb-0f9968bac445"}}
deepgram: transcript received
socket: transcript sent to client
deepgram: packet received

 Time 2023-12-07T22:47:19.978Z

 Transcription ==>  {"type":"Results","channel_index":[0,1],"duration":2.0600004,"start":15.44,"is_final":false,"speech_final":false,"channel":{"alternatives":[{"transcript":"","confidence":0,"words":[]}]},"metadata":{"request_id":"49397075-7dfb-4848-8419-190bc3d0d9b9","model_info":{"name":"2-ea-nova","version":"2023-09-18.22988","arch":"nova-2"},"model_uuid":"79dd3ac5-776e-4fcf-b4fb-0f9968bac445"}}
deepgram: transcript received
socket: transcript sent to client
deepgram: packet received

 Time 2023-12-07T22:47:20.865Z

 Transcription ==>  {"type":"Results","channel_index":[0,1],"duration":3.0600004,"start":15.44,"is_final":false,"speech_final":false,"channel":{"alternatives":[{"transcript":"","confidence":0,"words":[]}]},"metadata":{"request_id":"49397075-7dfb-4848-8419-190bc3d0d9b9","model_info":{"name":"2-ea-nova","version":"2023-09-18.22988","arch":"nova-2"},"model_uuid":"79dd3ac5-776e-4fcf-b4fb-0f9968bac445"}}
deepgram: transcript received
socket: transcript sent to client
deepgram: packet received

 Time 2023-12-07T22:47:21.845Z

 Transcription ==>  {"type":"Results","channel_index":[0,1],"duration":4.0600004,"start":15.44,"is_final":false,"speech_final":false,"channel":{"alternatives":[{"transcript":"","confidence":0,"words":[]}]},"metadata":{"request_id":"49397075-7dfb-4848-8419-190bc3d0d9b9","model_info":{"name":"2-ea-nova","version":"2023-09-18.22988","arch":"nova-2"},"model_uuid":"79dd3ac5-776e-4fcf-b4fb-0f9968bac445"}}
deepgram: transcript received
socket: transcript sent to client
deepgram: packet received

 Time 2023-12-07T22:47:22.817Z

 Transcription ==>  {"type":"Results","channel_index":[0,1],"duration":3.1400003,"start":15.44,"is_final":true,"speech_final":false,"channel":{"alternatives":[{"transcript":"","confidence":0,"words":[]}]},"metadata":{"request_id":"49397075-7dfb-4848-8419-190bc3d0d9b9","model_info":{"name":"2-ea-nova","version":"2023-09-18.22988","arch":"nova-2"},"model_uuid":"79dd3ac5-776e-4fcf-b4fb-0f9968bac445"}}
deepgram: transcript received
socket: transcript sent to client
deepgram: packet received

 Time 2023-12-07T22:47:22.821Z

 Transcription ==>  {"type":"Results","channel_index":[0,1],"duration":1.8600006,"start":18.58,"is_final":false,"speech_final":false,"channel":{"alternatives":[{"transcript":"","confidence":0,"words":[]}]},"metadata":{"request_id":"49397075-7dfb-4848-8419-190bc3d0d9b9","model_info":{"name":"2-ea-nova","version":"2023-09-18.22988","arch":"nova-2"},"model_uuid":"79dd3ac5-776e-4fcf-b4fb-0f9968bac445"}}
deepgram: transcript received
socket: transcript sent to client
deepgram: packet received

 Time 2023-12-07T22:47:23.799Z

 Transcription ==>  {"type":"Results","channel_index":[0,1],"duration":2.8799992,"start":18.58,"is_final":false,"speech_final":false,"channel":{"alternatives":[{"transcript":"This","confidence":0.9923455,"words":[{"word":"this","start":20.9,"end":21.4,"confidence":0.9923455,"punctuated_word":"This"}]}]},"metadata":{"request_id":"49397075-7dfb-4848-8419-190bc3d0d9b9","model_info":{"name":"2-ea-nova","version":"2023-09-18.22988","arch":"nova-2"},"model_uuid":"79dd3ac5-776e-4fcf-b4fb-0f9968bac445"}}
deepgram: transcript received
socket: transcript sent to client
deepgram: packet received

 Time 2023-12-07T22:47:24.678Z

 Transcription ==>  {"type":"Results","channel_index":[0,1],"duration":3.7199993,"start":18.58,"is_final":true,"speech_final":true,"channel":{"alternatives":[{"transcript":"This is","confidence":0.9850814,"words":[{"word":"this","start":20.98,"end":21.3,"confidence":0.9850814,"punctuated_word":"This"},{"word":"is","start":21.3,"end":21.8,"confidence":0.5317576,"punctuated_word":"is"}]}]},"metadata":{"request_id":"49397075-7dfb-4848-8419-190bc3d0d9b9","model_info":{"name":"2-ea-nova","version":"2023-09-18.22988","arch":"nova-2"},"model_uuid":"79dd3ac5-776e-4fcf-b4fb-0f9968bac445"}}
deepgram: transcript received
socket: transcript sent to client
deepgram: packet received

 Time 2023-12-07T22:47:25.679Z

 Transcription ==>  {"type":"Results","channel_index":[0,1],"duration":1,"start":22.3,"is_final":false,"speech_final":false,"channel":{"alternatives":[{"transcript":"","confidence":0,"words":[]}]},"metadata":{"request_id":"49397075-7dfb-4848-8419-190bc3d0d9b9","model_info":{"name":"2-ea-nova","version":"2023-09-18.22988","arch":"nova-2"},"model_uuid":"79dd3ac5-776e-4fcf-b4fb-0f9968bac445"}}
deepgram: transcript received
socket: transcript sent to client
deepgram: packet received

 Time 2023-12-07T22:47:26.673Z

 Transcription ==>  {"type":"Results","channel_index":[0,1],"duration":2,"start":22.3,"is_final":false,"speech_final":false,"channel":{"alternatives":[{"transcript":"","confidence":0,"words":[]}]},"metadata":{"request_id":"49397075-7dfb-4848-8419-190bc3d0d9b9","model_info":{"name":"2-ea-nova","version":"2023-09-18.22988","arch":"nova-2"},"model_uuid":"79dd3ac5-776e-4fcf-b4fb-0f9968bac445"}}
deepgram: transcript received
socket: transcript sent to client
deepgram: packet received

 Time 2023-12-07T22:47:27.672Z

 Transcription ==>  {"type":"Results","channel_index":[0,1],"duration":3,"start":22.3,"is_final":false,"speech_final":false,"channel":{"alternatives":[{"transcript":"","confidence":0,"words":[]}]},"metadata":{"request_id":"49397075-7dfb-4848-8419-190bc3d0d9b9","model_info":{"name":"2-ea-nova","version":"2023-09-18.22988","arch":"nova-2"},"model_uuid":"79dd3ac5-776e-4fcf-b4fb-0f9968bac445"}}
deepgram: transcript received
socket: transcript sent to client
deepgram: packet received

 Time 2023-12-07T22:47:27.681Z

 Transcription ==>  {"type":"UtteranceEnd"}
deepgram: utterence end

**Expected Transcript:- This is pickup**

**Can you guys please address this issue why is this happening?**

**Comment by team-deepgram** [comment]
Thanks for asking your question about Deepgram! If you didn't already include it in your post, please be sure to add as much detail as possible so we can assist you efficiently, such as:
- The `request_id` if you have a question about your requests or transcription responses.
- The features you used or the full `api.deepgram.com` URL you sent your request to, including parameters.
- Any code snippets you can share.

**Reply by aawezk786** [reply]
Hi team, I am waiting for reply

**Comment by jkroll-deepgram** [comment]
Hi @aawezk786, this can occur sometimes. How often are you observing it happening?

I see you're using our `nova-2-ea` model. We have a new GA `nova-2` model that has some improvements over the EA model. You could give that a try as well.

**Reply by aawezk786** [reply]
Since day one after integrating.

**Reply by aawezk786** [reply]
@jkroll-deepgram Can you please provide me the solution of above.I switched on nova-2 and much better perfomance i am getting but still there are some transcription cutoff why is this happening?
@team-deepgram

**Reply by cesmoak** [reply]
I'm seeing this behavior, as well. The last word or two is getting dropped from transcripts. I'm seeing it ~20% of the time using the nova-2 model. Here are a few example request ids:
- 1b25d239-2d9f-4bd1-a16b-09fd3a7088c0
- f44339f2-0e2d-4469-b13f-43e2bb9ab9d9
- eb04bc24-f49e-402e-a6c8-1548ab797a86
- a568843a-ea8b-41a6-b789-f8fab9edbc64

This is the code snippet I'm using:
```
const { result, error } = await deepgram.listen.prerecorded.transcribeFile(
    fs.readFileSync(pathToFile),
    {smart_format: true, model: "nova-2", language: "en-US"},
);
```

**Reply by aawezk786** [reply]
I want live streaming.How can i acheive that? @cesmoak  @jkroll-deepgram  @team-deepgram

**Reply by aawezk786** [reply]
@jkroll-deepgram Can you please provide solution for above question?

**Reply by aawezk786** [reply]
@team-deepgram @jkroll-deepgram waiting for solution

**Reply by jkroll-deepgram** [reply]
Hi @aawezk786, we are continuing to investigate improvements to our Nova-2 model to reduce the frequency of the final words being cut off from transcripts. If this is a blocking issue for you, we also have Enhanced and Base models with different architectures that do not exhibit this behavior (i.e. instead of `model=nova-2`, use `model=enhanced` or `model=base`). The tradeoff is that Nova-2 offers the highest overall accuracy.

For live-streaming, you can try our [Quickstart](https://developers.deepgram.com/docs/getting-started-with-live-streaming-audio) which offers code examples in Python, Javascript, C#, and Go, and we also offer an open-source [Streaming Test Suite](https://github.com/deepgram/streaming-test-suite/tree/main) which is written in Python, that demonstrates streaming files and from a local microphone.

**Reply by jjmaldonis** [reply]
Hey @aawezk786, all Nova 2 models have been updated with a fix for missing words at the end of transcripts. Make sure you're using `nova-2`, `nova-2-phonecall`, or one of the other Nova 2 models (e.g. `meeting`). Let us know if you continue to see missing words in transcripts.

**Reply by aawezk786** [reply]
@jjmaldonis nova-2,nova-2-conversationalai,nova is not working
language: "en-US",
      punctuate: false,
      encoding: "linear16",
      sample_rate: 16000,
      numerals: true,
      filler_words: false,
      vad_events: true,
      // endpointing: 500, //It also contains boolean value
      utterance_end_ms: 1400, //silence detection timeout
      interim_results: true,
      smart_format: true,
      model: process.env.DEEPGRAM_MODEL, //nova-2-conversationalai
      replace: hubConstants.FIND_AND_REPLACE,
      keywords: hubConstants.keywords_boost,
      
      but still i can see the transcription cut off 
      which model is best for transcription non cut off also i can use utterance_end_ms 
      
      the actual issue is i am getting utterence end message early although on call i was not stop

**Reply by aawezk786** [reply]
@team-deepgram @jkroll-deepgram Please provide solution its urgent now

**Reply by jkroll-deepgram** [reply]
Hi @aawezk786, I understand you're getting an UtteranceEnd message earlier than you'd expect. Since you are also using Endpointing, you can consider combining them both for end-of-speech detection. We have a guide on [End of Speech Detection While Live Streaming](https://developers.deepgram.com/docs/understanding-end-of-speech-detection#using-utteranceend-and-endpointing) that specifically addresses the logic of to combine them.

**Reply by aawezk786** [reply]
@jkroll-deepgram Can you provide me a code sample?

**Reply by jkroll-deepgram** [reply]
@aawezk786 Please see the link to "Download our latency Python example script" in the [Calculating Transcription Latency](https://developers.deepgram.com/docs/measuring-streaming-latency#calculating-transcription-latency) section.

**Reply by aawezk786** [reply]
JUST FYI endpointing is commented since that feature is also not working for me

**Reply by aawezk786** [reply]
@jkroll-deepgram  didnt get it how to calculate end statement using nova-2-conversationai model with using below config

language: "en-US",
punctuate: false,
encoding: "linear16",
sample_rate: 16000,
numerals: true,
filler_words: false,
vad_events: true,
utterance_end_ms: 1400, //silence detection timeout
interim_results: true,
smart_format: true,
model: process.env.DEEPGRAM_MODEL, //nova-2-conversationalai
replace: hubConstants.FIND_AND_REPLACE,
keywords: hubConstants.keywords_boost,

**Reply by jkroll-deepgram** [reply]
Here is the script we provide, with your parameters added. I made a small tweak since you're using SpeechStarted and UtteranceEnd events, to skip those for the latency calculations, as they're not transcription events so they don't have a `start` and `duration`. You can run the script with the command: `python latency.py -k <Deepgram API key> <path to audio file>`

```
""" Uses the real-time endpoint to calculate total latency.
    Python version: 3.6+
    Dependencies (use `pip install X` to install a dependency):
      - websockets
    Usage:
      python latency.py -k 'YOUR_DEEPGRAM_API_KEY' /path/to/audio.wav
    Limitations:
      - Only parses signed, 16-bit little-endian encoded WAV files.
    Notes:
      - Need a file? Grab one here: http://static.deepgram.com/examples/interview_speech-analytics.wav
"""

import argparse
import asyncio
import base64
import json
import sys
import wave
import websockets

# Mimic sending a real-time stream by sending this many seconds of audio at a time.
REALTIME_RESOLUTION = 0.020

async def run(data, key, channels, sample_width, sample_rate):
    """ Connect to the Deepgram real-time streaming endpoint, stream the data
        in real-time, and print out the responses from the server.

        This uses a pre-recorded file as an example. It mimics a real-time
        connection by sending `REALTIME_RESOLUTION` seconds of audio every
        `REALTIME_RESOLUTION` seconds of wall-clock time.

        This is a toy example, since it uses a pre-recorded file. In a real use
        case, you'd be streaming the audio stream itself. If you actually have
        pre-recorded audio, use Deepgram's pre-recorded mode instead.
    """
    # How many bytes are contained in one second of audio.
    byte_rate = sample_width * sample_rate * channels

    audio_cursor = 0.

    # Connect to the real-time streaming endpoint, attaching our API key.
    async with websockets.connect(
        f'wss://api.deepgram.com/v1/listen?channels={channels}&sample_rate={sample_rate}&encoding=linear16&interim_results=true&language=en-US&model=nova-2-conversationalai&smart_format=true&numerals=true&vad_events=true&utterance_end_ms=1400',
        extra_headers={
            'Authorization': 'Token {}'.format(key)
        }
    ) as ws:
        async def sender(ws):
            """ Sends the data, mimicking a real-time connection.
            """
            nonlocal data, audio_cursor
            try:
                total = len(data)
                while len(data):
                    # How many bytes are in `REALTIME_RESOLUTION` seconds of audio?
                    i = int(byte_rate * REALTIME_RESOLUTION)
                    chunk, data = data[:i], data[i:]
                    # Send the data
                    await ws.send(chunk)
                    # Move the audio cursor
                    audio_cursor += REALTIME_RESOLUTION
                    # Mimic real-time by waiting `REALTIME_RESOLUTION` seconds
                    # before the next packet.
                    await asyncio.sleep(REALTIME_RESOLUTION)

                # A CloseStream message tells Deepgram that no more audio
                # will be sent. Deepgram will close the connection once all
                # audio has finished processing.
                await ws.send(json.dumps({
                    "type": "CloseStream"
                }))
            except Exception as e:
                print(f'Error while sending: {e}')
                raise

        async def receiver(ws):
            """ Print out the messages received from the server.
            """
            nonlocal audio_cursor
            transcript_cursor = 0.
            min_latency = 0
            max_latency = 0
            avg_latency_num = 0
            avg_latency_den = 0
            async for msg in ws:
                msg = json.loads(msg)
                if 'request_id' in msg:
                    # This is the final metadata message. It gets sent as the
                    # very last message by Deepgram during a clean shutdown.
                    # There is no transcript in it.
                    continue
                if msg['type'] in ['SpeechStarted', 'UtteranceEnd']:
                    continue
                if msg['is_final']:
                    continue
                # Just a moment ago -- right before we received this message --
                # is when the latency was at its worse. So measure max latency
                # and then update the transcript_cursor
                cur_max_latency = audio_cursor - transcript_cursor

                transcript_cursor = msg['start'] + msg['duration']

                # Since we just received a message, latency is currently at its
                # best.
                cur_min_latency = audio_cursor - transcript_cursor

                # The average latency (as would be measured by a constant probe
                # or by a sampling profiler) is mathematically equivalent to a
                # weighted sum.
                avg_latency_num += (cur_min_latency + cur_max_latency) / 2 * msg['duration']
                avg_latency_den += msg['duration']

                # Update global max/min latencies.
                max_latency = max(max_latency or 0, cur_max_latency)
                min_latency = min(min_latency or 10**6, cur_min_latency)

                print(f'Measuring... Audio cursor = {audio_cursor:.3f}, Transcript cursor = {transcript_cursor:.3f}')

            print(f'Min latency: {min_latency:.3f}')
            print(f'Avg latency: {avg_latency_num / (avg_latency_den or 1):.3f}')
            print(f'Max latency: {max_latency:.3f}')

        await asyncio.wait([
            asyncio.ensure_future(sender(ws)),
            asyncio.ensure_future(receiver(ws))
        ])

###############################################################################
def parse_args():
    """ Parses the command-line arguments.
    """
    parser = argparse.ArgumentParser(description='Submits data to the real-time streaming endpoint.')
    parser.add_argument('-k', '--key', required=True, help='YOUR_DEEPGRAM_API_KEY (authorization)')
    parser.add_argument('input', help='Input file.')
    return parser.parse_args()

###############################################################################
def main():
    """ Entrypoint for the example."
    """
    # Parse the command-line arguments.
    args = parse_args()

    # Open the audio file.
    with wave.open(args.input, 'rb') as fh:
        (channels, sample_width, sample_rate, num_samples, _, _) = fh.getparams()
        assert sample_width == 2, 'WAV data must be 16-bit.'
        data = fh.readframes(num_samples)
    print(f'Channels = {channels}, Sample Rate = {sample_rate} Hz, Sample width = {sample_width} bytes, Size = {len(data)} bytes', file=sys.stderr)

    # Run the example.
    asyncio.get_event_loop().run_until_complete(run(data, args.key, channels, sample_width, sample_rate))

###############################################################################
if __name__ == '__main__':
    sys.exit(main() or 0)
```

Here is the output I got from streaming 20 seconds of audio. My audio happens to be 48000 Hz rather than your 16000 Hz.
```
Channels = 1, Sample Rate = 48000 Hz, Sample width = 2 bytes, Size = 1890176 bytes
Measuring... Audio cursor = 1.200, Transcript cursor = 1.000
Measuring... Audio cursor = 2.140, Transcript cursor = 2.000
Measuring... Audio cursor = 3.180, Transcript cursor = 3.000
Measuring... Audio cursor = 4.200, Transcript cursor = 4.000
Measuring... Audio cursor = 5.780, Transcript cursor = 5.580
Measuring... Audio cursor = 7.680, Transcript cursor = 7.520
Measuring... Audio cursor = 8.680, Transcript cursor = 8.520
Measuring... Audio cursor = 9.700, Transcript cursor = 9.520
Measuring... Audio cursor = 10.700, Transcript cursor = 10.520
Measuring... Audio cursor = 12.720, Transcript cursor = 12.500
Measuring... Audio cursor = 13.700, Transcript cursor = 13.500
Measuring... Audio cursor = 15.720, Transcript cursor = 15.560
Measuring... Audio cursor = 16.740, Transcript cursor = 16.560
Measuring... Audio cursor = 17.720, Transcript cursor = 17.560
Measuring... Audio cursor = 18.700, Transcript cursor = 18.560
Min latency: 0.140
Avg latency: 0.727
Max latency: 2.220
```
